### DISCUSSION TITLE
Sleep/Lightsleep with Pico 2

### DISCUSSION THREAD

**Discussion by ssotheremail** [micropython/micropython#16361] [RP2040 / Pico]
Firstly thanks for creating and maintaining Micropython - its a fantastic resource.

Forgive me for opening this as a new issue, and its length, but this matter has been mentioned a lot of times in different ways and I could not find one to reply to that covered it comprehensively.  The code below has been run against several Micropython versions as you can see. The key ones here are:
v1.24.1 on 2024-11-29 Raspberry Pi Pico2 with RP2350
v1.25.0-preview.73.g406bccc75 on 2024-12-03; Raspberry Pi Pico2 with RP2350

There seem to be 3 main issues which are causing difficulties.

1. I have stopped using usb and print as lightsleep messes it up and its even harder to work out what's happening. In production its not too bad as if you detect usb connected you can choose to print and use sleep delays instead of light sleep for debugging, or if not connected, use light sleep.
2.  The most insidious problem is demonstrated in Pass 2 where the time.sleep(2.5) seems to stop the following lightsleep(2500) from providing any delay at all. Initially I didn't believe this was a real effect, but it seems to be.
3. It seems time.time() does not advance during lightsleep so my output file says its finished in about 70sec whereas you would expect 5x5x5 to be 125sec. Even more confusingly it actually finishes in ~115sec because of pass 2.

I have included 1 output file below as an example. If anyone wishes to take this further please reply here. 

```
import  machine
import time

# machine.freq(48000000) #its on battery see what happens when we set the mcu to 48mHz
#file = open("v1.24.1 on 2024-11-29 Raspberry Pi Pico2 with RP2350", "w")
#file = open("v1.24.1 on 2024-11-29; Raspberry Pi Pico with RP2040", "w")
#file = open("Pimoroni feature psram-and-wifi, pico2_rp2350 v0.0.9 on 2024-11-06; Raspberry Pi Pico 2 with RP2350", "w")
#file = open("Pimoroni feature psram-and-wifi pico2 v0.0.11 on 2024-11-26; Raspberry Pi Pico2 with RP2350", "w")
#file = open("MicroPython v1.25.0-preview.72.g2e796d6c3 on 2024-11-30; Raspberry Pi Pico2 with RP2350", "w")
file = open("v1.25.0-preview.73.g406bccc75 on 2024-12-03; Raspberry Pi Pico2 with RP2350", "w")

led = machine.Pin("LED", machine.Pin.OUT)
led.off()

start_time = time.time()

#while True:
for y in range(0,5):
    #print("Loop instance with delay provided by for loops:",y,time.time()-start_time)
    file.write("Pass 1: Loop instance with for loops:" + "{:5d} {:5d}".format(y,time.time()-start_time)+"\n")
    file.flush()
    for x in range(0,400000): led.off()
    for x in range(0,400000): led.on()
    
for y in range(0,5):
    #print("Loop instance with delay provided by machine.lightsleep",y,time.time()-start_time)
    file.write("Pass 2: Loop instance with time.sleep and machine.lightsleep" + "{:5d} {:5d}".format(y,time.time()-start_time)+"\n")
    file.flush()
    led.off()
    #machine.lightsleep(2500)
    time.sleep(2.5)
    led.on()
    machine.lightsleep(2500) 
    
    
for y in range(0,5):
    #print("Loop instance with delay provided by machine.lightsleep",y,time.time()-start_time)
    file.write("Pass 3: Loop instance with machine.lightsleep" + "{:5d} {:5d}".format(y,time.time()-start_time)+"\n")
    file.flush()  
    led.off()
    #time.sleep(1.0)
    machine.lightsleep(2500)import  machine
import time

# machine.freq(48000000) #its on battery see what happens when we set the mcu to 48mHz
#file = open("v1.24.1 on 2024-11-29 Raspberry Pi Pico2 with RP2350", "w")
#file = open("v1.24.1 on 2024-11-29; Raspberry Pi Pico with RP2040", "w")
#file = open("Pimoroni feature psram-and-wifi, pico2_rp2350 v0.0.9 on 2024-11-06; Raspberry Pi Pico 2 with RP2350", "w")
#file = open("Pimoroni feature psram-and-wifi pico2 v0.0.11 on 2024-11-26; Raspberry Pi Pico2 with RP2350", "w")
#file = open("MicroPython v1.25.0-preview.72.g2e796d6c3 on 2024-11-30; Raspberry Pi Pico2 with RP2350", "w")
file = open("v1.25.0-preview.73.g406bccc75 on 2024-12-03; Raspberry Pi Pico2 with RP2350", "w")

led = machine.Pin("LED", machine.Pin.OUT)
led.off()

start_time = time.time()

#while True:
for y in range(0,5):
    #print("Loop instance with delay provided by for loops:",y,time.time()-start_time)
    file.write("Pass 1: Loop instance with for loops:" + "{:5d} {:5d}".format(y,time.time()-start_time)+"\n")
    file.flush()
    for x in range(0,400000): led.off()
    for x in range(0,400000): led.on()
    
for y in range(0,5):
    #print("Loop instance with delay provided by machine.lightsleep",y,time.time()-start_time)
    file.write("Pass 2: Loop instance with time.sleep and machine.lightsleep" + "{:5d} {:5d}".format(y,time.time()-start_time)+"\n")
    file.flush()
    led.off()
    #machine.lightsleep(2500)
    time.sleep(2.5)
    led.on()
    machine.lightsleep(2500) 
    
    
for y in range(0,5):
    #print("Loop instance with delay provided by machine.lightsleep",y,time.time()-start_time)
    file.write("Pass 3: Loop instance with machine.lightsleep" + "{:5d} {:5d}".format(y,time.time()-start_time)+"\n")
    file.flush()  
    led.off()
    #time.sleep(1.0)
    machine.lightsleep(2500)
    led.on()
    machine.lightsleep(2500)v1.25.0-preview.73.g406bccc75 on 2024-12-03; Raspberry Pi Pico2 with RP2350
    
for y in range(0,5):
    #print("Loop instance with delay provided by time.sleep",y,time.time()-start_time)
    file.write("Pass 4: Loop instance with time.sleep" + "{:5d} {:5d}".format(y,time.time()-start_time)+"\n")
    file.flush()
    led.off()
    time.sleep(2.5)
    led.on()
    time.sleep(2.5)
    
    
for y in range(0,5):
    #print("Loop instance with delay provided by machine.lightsleep",y,time.time()-start_time)
    file.write("Pass 5: Loop instance with machine.lightsleep" + "{:5d} {:5d}".format(y,time.time()-start_time)+"\n")
    file.flush()  
    led.off()
    #time.sleep(1.0)
    machine.lightsleep(2500)
    led.on()
    machine.lightsleep(2500)

file.write("Expected run time ~ "+"{:5d}".format(24+(4*5*5))+" seconds Actual 115 sec\n")

file.close()
    led.on()
    machine.lightsleep(2500)
    
for y in range(0,5):
    #print("Loop instance with delay provided by time.sleep",y,time.time()-start_time)
    file.write("Pass 4: Loop instance with time.sleep" + "{:5d} {:5d}".format(y,time.time()-start_time)+"\n")
    file.flush()
    led.off()
    time.sleep(2.5)
    led.on()
    time.sleep(2.5)
    
    
for y in range(0,5):
    #print("Loop instance with delay provided by machine.lightsleep",y,time.time()-start_time)
    file.write("Pass 5: Loop instance with machine.lightsleep" + "{:5d} {:5d}".format(y,time.time()-start_time)+"\n")
    file.flush()  
    led.off()
    #time.sleep(1.0)
    machine.lightsleep(2500)
    led.on()
    machine.lightsleep(2500)

file.write("Expected run time ~ "+"{:5d}".format(24+(4*5*5))+" seconds Actual 115 sec\n")

file.close()
```

output file:v1.25.0-preview.73.g406bccc75 on 2024-12-03; Raspberry Pi Pico2 with RP2350
Pass 1: Loop instance with for loops:    0     0
Pass 1: Loop instance with for loops:    1     5
Pass 1: Loop instance with for loops:    2     9
Pass 1: Loop instance with for loops:    3    14
Pass 1: Loop instance with for loops:    4    19
Pass 2: Loop instance with time.sleep and machine.lightsleep    0    23
Pass 2: Loop instance with time.sleep and machine.lightsleep    1    26
Pass 2: Loop instance with time.sleep and machine.lightsleep    2    28
Pass 2: Loop instance with time.sleep and machine.lightsleep    3    31
Pass 2: Loop instance with time.sleep and machine.lightsleep    4    34
Pass 3: Loop instance with machine.lightsleep    0    36
Pass 3: Loop instance with machine.lightsleep    1    36
Pass 3: Loop instance with machine.lightsleep    2    36
Pass 3: Loop instance with machine.lightsleep    3    36
Pass 3: Loop instance with machine.lightsleep    4    36
Pass 4: Loop instance with time.sleep    0    36
Pass 4: Loop instance with time.sleep    1    41
Pass 4: Loop instance with time.sleep    2    46
Pass 4: Loop instance with time.sleep    3    51
Pass 4: Loop instance with time.sleep    4    56
Pass 5: Loop instance with machine.lightsleep    0    61
Pass 5: Loop instance with machine.lightsleep    1    61
Pass 5: Loop instance with machine.lightsleep    2    61
Pass 5: Loop instance with machine.lightsleep    3    61
Pass 5: Loop instance with machine.lightsleep    4    61
Expected run time ~ 124 seconds Actual 115 sec

**Comment by scruss** [comment]
What are you trying to do? Verify that lightsleep() is accurate?

There have been problems with lightsleep in the past, but I don't know how many of them have been addressed.

Incidentally, you can pick up the version/hardware from os.uname():

```
>>> import os
>>> os.uname()
(sysname='rp2', nodename='rp2', release='1.24.0', version='v1.24.0 on 2024-10-25 (GNU 13.2.0 MinSizeRel)', machine='Raspberry Pi Pico with RP2040')
```

**Reply by ssotheremail** [reply]
Thanks for the reponse (and useful tip).
Perhaps some context would help. One of my pico projects logs 3 temperatures every minute and writes them to a file. It runs off batteries for safety/other reasons. I used print statements to help get it going. When I started to use lightsleep to enhance the battery life I noticed I couldn't really use print any more. 
I needed to buy some more picos at about the time the pico 2 came out, saw the documentation for lower power modes and thought I would give them a go. I was a bit surprised that there seemed to be so much uncertainty about the performance of sleeping and in some cases it was getting worse. E.g. Pimoroni 0.0.9 for the 2350 broadly seems to do the same as "official" 1.24.1, but in the latest preview 0.0.11 lightsleep(2500) never seems to return at all (I will comment on that to them). I experimented briefly with someone's example with lightsleep() and a timer to wake it up and that didn't seem to wake at all. 

My thought in starting this comment was to try to help to pin down, in a simple way, the things that are currently "wrong" with lightsleep in the hope that 1.25 "official" and Pimoroni's equivalent release of Micropython will work much more suitably for me (and presumably at least some other people!) As far as I can tell from github (not my strength) progress at present seems uncertain.

Hope this clarifies. 

thanks again

**Reply by scruss** [reply]
> there seemed to be so much uncertainty about the performance of sleeping

What do you mean? Current draw? Timing?

You can't print during/near lightsleep. Print means active USB, so no lightsleep (or more likely, lightsleep confuses your serial connection)

**Reply by ssotheremail** [reply]
Agreed about not using print with lightsleep, that became clear by experiment!
 What I meant by uncertainty is that it appears (to me) that there have been bugs in the 2.0.0 sdk, regressions in Micropython, unstable code and for example the ongoing debate in:
github.com/micropython/micropython/pull/16313
on how these issues can be fixed. It appears to raise the possibility of yet another bug in the current Pico C Sdk.
I am not for a moment intending to be critical of people doing the brilliant work going on with these devices, only trying to provide evidence to help fix it. 
Anyhow thanks for your time I will leave you in peace!

**Comment by mungewell** [comment]
I am successfully using `lightsleep()` in my project, but I have hack-around quiet a bit under the hood. If you are willing to rebuild your own version of micropython you can check out:
https://github.com/micropython/micropython/pull/16183

I am doing this to leaving particular clocks (and 'parents') running during `lightsleep()`. This means that I can keep the PIO blocks running whilst sleeping, for me this is better (power wise) than keeping the CPU alive. 

I also found issues with UART being functional after a sleep, so I keep the peripheral clock running (I need to update/resubmit the patch for that one).

**Comment by CrabbyPete** [comment]
I have code that worked great in 1.23  on a Pico  ( not 2)  and in 1.24 no longer works at all.   I am using it to connect to a modem that is in deep sleep and I need to sync the  two [https://github.com/CrabbyPete/watchible/blob/1a3adbdcdd1f291458ea1193a4a8e1a1a67e1057/software/python/main.py#L756](url) Did something change?

**Reply by CrabbyPete** [reply]
(sysname='rp2', nodename='rp2', release='1.24.1', version='v1.24.1 on 2024-11-29 (GNU 13.2.0 MinSizeRel)', machine='Raspberry Pi Pico with RP2040')

**Reply by scruss** [reply]
a lot has changed in lightsleep, seemingly mostly to allow for the wifi to sleep gracefully and any USB devices to be dealt with. I may not be linking to the right diff between v1.23.0 and v1.24.1 here: [Comparing v1.23.0...v1.24.1 · micropython/micropython](https://github.com/micropython/micropython/compare/v1.23.0...v1.24.1#diff-c2265d7118c08ee4957f498973f1d498fba55bf3909f924448c657d7419ee562L108) ... you're looking for the file ports/rp2/modmachine.c in all of this, if the link doesn't do the thing.

I'm surprised that you're seeing this on a regular non-wireless RP2040 Pico. I can only make the example code on the issue [lightsleep on Pico W regression for 1.24.0 · Issue #16181 · micropython/micropython](https://github.com/micropython/micropython/issues/16181) fail on a Pico W, not on a Pico

**Reply by CrabbyPete** [reply]
I tried again and same result. I don't know what's written on the chip the board says Raspberry Pi Pico 2020.  It fails every time. I wrote a simple program and tried to run it. 

`from time import sleep, gmtime
from machine import lightsleep

while True:
    lightsleep(3600)
    print (f'{gmtime()}')
    sleep(.1)`

and I get
(2024, 12, 10, 16, 25, 48, 1, 345)
(2024, 12, 10, 16, 25, 48, 1, 345)
(2024, 12, 10, 16, 25, 48, 1, 345)
(2024, 12, 10, 16, 25, 48, 1, 345)

following this post [https://github.com/orgs/micropython/discussions/12573#discussioncomment-11525311](url)

**Reply by ssotheremail** [reply]
I have done a few more experiments. I hope the following may help, concentrating on Pico 2 and Pico 2w. I have not used print in order to simplify things.

I only have access to 2 preview builds for Pico 2 W:
MicroPython v1.25.0-preview.49.g0625f07ad.dirty on 2024-11-21; Raspberry Pi Pico 2 W with RP2350: 1st lightsleep(2500) never wakes up. program stalls indefinitely.
Micropython v1.25.0-preview.88.g2d4b02b4d(on 2024-12-08)_Pico2W: No perceptible length of sleep with lightsleep(2500) will try a better timing method in case the actual length of sleep is a clue.
 
On the Pico 2 with several previews of 1.25.0 up to 81 lightsleep(2500) provided an accurate sleep time as long as you don't mix it with time.sleep in the loop.  See pass 1 in attached Sample_file.txt . If you do then  lightsleep(2500) doesn't seem to sleep. 

I have added time.ticks_ms between the loops and the time calculated from time.ticks_ms is the very close to the real execution time measured manually on a stopwatch. However time.time and presumably gmtime etc come up with a smaller execution time as it appears that time.time stop stops incrementing during lightsleep.  Is this intended? I didn't find it documented. 

E.g: Expected run time ~50 seconds. 
Elapsed Measured by Pico time.time:  13 sec.  Actual with stopwatch 38 sec, with time.ticks_ms 37758

The "missing" time from 50 seconds is made up by 12.5 seconds in pass 1 that should have been spent sleeping but wasn't sleeping, leaving a possible 37.5 seconds (as correctly recorded by ticks_ms) and 25 seconds in pass 2 not recorded by time.time.

When the program is run with Thonny over usb the file closes properly. I tried a couple of times to run it as main.py when powered by a usb charger and it didn't finish writing the file and/or didn't close it. I will look into that tomorrow as it may mean there is a difference when usb is or is not connected.

If it would help I could run against Pico  or Pico W too.

[Sample_file.txt](https://github.com/user-attachments/files/18086868/Sample_file.txt)

My current test program is:

```
import machine
import time


file = open("MicroPython v1.25.0-preview.81.g2c80d3699 on 2024-12-06; Raspberry Pi Pico2 with RP2350", "w")

led = machine.Pin("LED", machine.Pin.OUT)
led.off()

start_time = time.time()
before_tick = time.ticks_ms()
after_tick = time.ticks_ms()
sleep_length = after_tick - before_tick
    
for y in range(0,5):
    #print("Loop instance with delay provided by machine.lightsleep",y,time.time()-start_time)
    file.write("Pass 1: Loop instance with time.sleep and machine.lightsleep" + "{:5d} {:5d}{:10d} ".format(y,time.time()-start_time,sleep_length)+"\n")
    file.flush()
    led.off()
    #machine.lightsleep(2500)
    time.sleep(2.5)
    led.on()
    machine.lightsleep(2500)
    
after_tick = time.ticks_ms()
sleep_length = after_tick - before_tick
    
for y in range(0,5):
    #print("Loop instance with delay provided by machine.lightsleep",y,time.time()-start_time)
    file.write("Pass 2: Loop instance with machine.lightsleep" + "{:5d} {:5d}{:10d} ".format(y,time.time()-start_time,sleep_length)+"\n")
    file.flush()  
    led.off()
    #time.sleep(1.0)
    machine.lightsleep(2500)
    led.on()
    machine.lightsleep(2500)
    
after_tick = time.ticks_ms()
sleep_length = after_tick - before_tick
    
file.write("Expected run time ~ "+"{:5d}".format((2*5*5))+" seconds. Elapsed Measured by Pico: "+"{:5d} {:5d}".format(time.time()-start_time,sleep_length) + "\n")

file.close()

```

**Comment by CrabbyPete** [comment]
I think the work by Carl here can help to figure this out. https://github.com/orgs/micropython/discussions/12573#discussioncomment-11514661
